tests/test_crud.py

```diff
@@ -127,61 +130,238 @@ def test_duplicated(self):
-            topic=datapoint_basis.topic,
+        datapoint_basis1.entity_id = None  # Ensure this triggers a failure
+        datapoint_basis1.entity_type = None  # Ensure this triggers a failure
+        datapoint_basis1.attribute_name = None  # Ensure this triggers a failure
+        update_data = datapoint_basis1.json()
+        response1 = requests.request("PUT", settings.GATEWAY_URL + "/data/" + object_id, headers=headers,
+                                     data=update_data)
+
+        # Log the request and response for debugging
+        print(f"Request data for update: {update_data}")
+        print(f"Response1 status code: {response1.status_code}")
+        print(f"Response1 content: {response1.text}")
+
+        # Ensure the response indicates failure
+        self.assertFalse(response1.ok)
+        self.assertEqual(response1.status_code, 400)
+        self.assertIn("entity_id, entity_type, and attribute_name cannot be null", response1.text)
+
+        # Perform a valid update with correct entity information
+        datapoint_basis_update = DatapointUpdate(
             entity_id=self.test_entity.id,
             entity_type=self.test_entity.type,
             attribute_name=self.test_entity.get_attribute_names().pop(),
         )
-        response2 = requests.request("PUT", settings.GATEWAY_URL + "/data/" + object_id,
-                                     headers=headers,
-                                     data=datapoint_basis_update.json()
-                                     )
+        response2 = requests.request("PUT", settings.GATEWAY_URL + "/data/" + object_id, headers=headers,
+                                     data=datapoint_basis_update.json())
         self.assertTrue(response2.ok)
 
-        # check update result
+        # Verify the update result
         datapoint_basis2: Datapoint = datapoint_basis.copy()
         datapoint_basis2.entity_id = self.test_entity.id
         datapoint_basis2.entity_type = self.test_entity.type
         datapoint_basis2.attribute_name = self.test_entity.get_attribute_names().pop()
-        # datapoint_basis2.matchDatapoint = True
         response = requests.request("GET", settings.GATEWAY_URL + "/data/" + object_id)
+
+        updated_datapoint = Datapoint(**json.loads(response.text))
+
+        self.assertEqual(
+            datapoint_basis2.entity_id,
+            updated_datapoint.entity_id
+        )
+        self.assertEqual(
+            datapoint_basis2.entity_type,
+            updated_datapoint.entity_type
+        )
         self.assertEqual(
-            datapoint_basis2.dict().pop("matchDatapoint"),
-            json.loads(response.text).pop("matchDatapoint")
+            datapoint_basis2.attribute_name,
+            updated_datapoint.attribute_name
+        )
+
+    def test_match_datapoints(self):
+        headers = {
+            'Accept': 'application/json'
+        }
+
+        # create matched datapoint
+        datapoint = Datapoint(
```

Comment:
You can actually reuse `self.matched_datapoint`, which is created in `test_init.py`. And actually I found that `dp:001` is not created. How would you expect the connection check can go through?

Reply:
fixed using self.matched_datapoint, the test case should pass now